File: 2021/11/05/CanBeTyped.java
/**
 * https://leetcode.com/problems/maximum-number-of-words-you-can-type/
 * @author Andrew Jarombek
 * @since 11/5/2021
 */

import java.util.HashSet;
import java.util.Set;

public class CanBeTyped {
    public int canBeTypedWords(String text, String brokenLetters) {
        int count = 0;

        Set<Character> brokenSet = new HashSet<>();

        for (char c: brokenLetters.toCharArray()) {
            brokenSet.add(c);
        }

        for (String word: text.split(" ")) {
            int index = 0;
            while (index < word.length() && !brokenSet.contains(word.charAt(index))) {
                index++;
            }

            if (index == word.length()) {
                count++;
            }
        }

        return count;
    }
}

File: 2022/10/04/canBeTyped.ts
/**
 * https://leetcode.com/problems/maximum-number-of-words-you-can-type/
 * @author Andrew Jarombek
 * @since 10/4/2022
 */

function canBeTypedWords(text: string, brokenLetters: string): number {
    let broken = new Set<string>();

    for (let c of brokenLetters.split("")) {
        broken.add(c);
    }

    let result = 0;

    for (let word of text.split(" ")) {
        let typed = true;
        let index = 0;

        while (index < word.length && typed) {
            if (broken.has(word.charAt(index))) {
                typed = false;
            } else {
                index++;
            }
        }

        if (typed)
            result++;
    }

    return result;
}
